src/com/ansorgit/plugins/bash/documentation/BashDocumentationProvider.kt

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ansorgit.plugins.bash.documentation
+
+import com.ansorgit.plugins.bash.lang.lexer.BashTokenTypes
+import com.ansorgit.plugins.bash.lang.psi.api.command.BashCommand
+import com.ansorgit.plugins.bash.lang.psi.api.command.BashGenericCommand
+import com.ansorgit.plugins.bash.lang.psi.api.vars.BashVar
+import com.ansorgit.plugins.bash.lang.psi.util.BashPsiUtils
+import com.intellij.lang.documentation.AbstractDocumentationProvider
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.util.SystemInfoRt
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiManager
+import com.intellij.psi.TokenType
+import org.apache.commons.lang.StringUtils
+
+/**
+ * This is the bash documentation provider. A documentation provider is called
+ * for PsiElements which implement the interface PsiReference.
+ * <br></br>
+ *
+ * @author jansorg
+ */
+class BashDocumentationProvider : AbstractDocumentationProvider() {
+    private val sourceList = mutableListOf(PsiElementCommentSource(), BashKeywordDocSource(), InternalCommandDocumentation(), ManpageDocSource())
+
+    init {
+        if (!SystemInfoRt.isWindows) {
+            //there is currently no support for the cygwin info command
+            sourceList.add(CachingDocumentationSource(SystemInfopageDocSource()))
+        }
+    }
+
+    override fun getCustomDocumentationElement(editor: Editor, file: PsiFile, contextElement: PsiElement?): PsiElement? {
+        if (contextElement is BashVar) {
+            return null
+        }
+
+        var psi = contextElement
+        if (psi == null) {
+            // probably invoked for element at end of file
+            val offset = editor.caretModel.offset - 1
+            if (offset >= 0) {
+                psi = file.findElementAt(offset)
+            }
+        }
+
+        val elementType = psi?.node?.elementType
+        if (psi != null && (elementType === BashTokenTypes.LINE_FEED || elementType === TokenType.WHITE_SPACE)) {
+            psi = psi.prevSibling
+        }
+
+        if (psi is BashCommand) {
+            psi = psi.commandElement()
+        }
+
+        return BashPsiUtils.findParent(psi, BashGenericCommand::class.java)
+    }
+
+    override fun getQuickNavigateInfo(element: PsiElement?, originalElement: PsiElement?): String? = null
+
+    override fun getUrlFor(element: PsiElement?, originalElement: PsiElement?): List<String>? {
+        for (source in sourceList) {
+            val url = source.documentationUrl(element, originalElement)
+            if (url != null && StringUtils.stripToNull(url) != null) {
+                return listOf(url)
```

Comment:
Shouldn't result of `stripToNull()` be returned instead? Otherwise it would be more efficient to use `StringUtils.isBlank()` in the condition.
(Also below in `generateDoc()`)